### PR TITLE
new function: strAppend(1+2, "foo")

### PR DESCRIPTION
**Pull request by timotheecour**
/cc @Araq 

* allows: `strAppend(arg1, arg2, arg3, ...)`
IMO `strAppend` is more readable in lots of cases, eg:
```nim
strAppend("foo", 1+2, 3==4, a[0])
"foo" & $(1+2) & $(3==4) & $a[0]  # too many symbols; it's hard to parse & and $ and easy to mistype (for me!)
```

### CODE REVIEW

**Comment by data-man** [comment]
Why in the new module? :-D

**Comment by timotheecour** [comment]
can you suggest which module? this function should be importable without pulling in tons of dependencies so I'd rather not put it in something that has many dependencies.
Likewise, adding it in strutils would add a dependency from strutils to macros.

**Comment by data-man** [comment]
I proposed to @Araq to create an official repository of snippets.
There are tons useful snippets in PRs (in third-party repositories too), issues, IRC, forum, etc.
And they are not easy to find, especially for beginners.
Also many snippets from tests can be adapted for this proposal.

**Comment by timotheecour** [comment]
I don't disagree with having an official repository of snippets; however, regardless of snippets repo, I'd still like this in stdlib so it can be used in most modules in stdlib.

**Comment by data-man** [comment]
Well, what's next?
```veryCoolProc``` in ```verycoolprocs.nim```?

**Comment by timotheecour** [comment]
I'm fine with an existing module but which one?

**Comment by data-man** [comment]
> which one?

No ideas. :)

**Comment by timotheecour** [comment]
see: BUG: "varargs[string, `$`]" calls `$` n^2 times instead of n times (n=len(varargs)) #8316

**Comment by yglukhov** [comment]
That bug affects only templates with varargs arguments, my sample (with procs) is not affected.

**Comment by dom96** [comment]
I don't understand the purpose of this macro. Can you explain?

**Comment by timotheecour** [comment]
PTAL:

@yglukhov 
* moved to macros.nim because it only depends on macros.nim and I want to avoid adding dependencies in existing modules for this particular PR (IIRC strutils doens't depend on macros.nim so wasn't such a good idea); on the other hand, importing this PR shouldn't bring more dependencies than strictly necessary so it could be used in more places in stdlib, hence it's logical to just put it in macros.nim

> how is it better than:
```nim
proc append(s: var string, args: varargs[string, `$`]) = for a in args: s.add(a)
proc join(args: varargs[string, `$`]): string = result.append(args)
```
efficiency: I benchmarked computing `let msg=strJoin("hello", i, n, [1,n], (i+n, i))` in a loop (100000 times):
* the version with `varargs[string, `$`]` takes `0.163s`
* the previous version of my PR takes `0.147s` (it didn't benefit from term rewriting though, not sure how to make it work without using parseExpr as I did in latest PR version)
* the latest version of my PR takes: `0.122s` (identical to calling directly `"hello" & $i & $n & $[1,n] & $(i+n, i)` ; this isn't suprising given the new implementation.

these are repeatable results.

@dom96 
> I don't understand the purpose of this macro. Can you explain?

clarified in top-level msg. boils down to:
* less visual noise due to excess symbols we often end up with:
```nim
strJoin("foo", 1+2, 3==4, a[0])
# vs
"foo" & $(1+2) & $(3==4) & $a[0]
```
* makes it easier to switch back and forth between echo statement and returning string: with strJoin you just replace `s/echo/strJoin/` ; with `&` you have to change a bunch of things, including adding `$` in places where arguments need stringification

**Comment by yglukhov** [comment]
> efficiency

@timotheecour is efficiency coming from preallocating the result string of sufficient capacity? If so, my example could be easily patched to account for that.

**Comment by timotheecour** [comment]
no, I'm not pre-allocating; it would be faster if I were (likewise for all approaches)
(and btw, I have another thread on that topic somewhere else; it's a different issue though)

```nim
proc test()=
  let n=100000
  var i=0
  benchmark("strAdd", n): # runs for n iters and reports timing
      let msg=strJoin("hello", i, n, [1,n], (i+n, i))
      i.inc
      msg.preventOptimizeOut # accesses some element to prevent optimizer from discarding
```

**Comment by data-man** [comment]
I dislike the name and thought this macro joins only strings (I know, because my English is bad). 
Maybe ```toString```? Or ```joinAsStr(ing)```?
And move it to `strformat`?

**Comment by LemonBoy** [comment]
@timotheecour I take you didn't see [this beautiful benchmark harness](https://github.com/LemonBoy/criterion.nim) yet :)

File: lib/core/macros.nim [hunk 1]
```diff
@@ -463,6 +463,24 @@ proc parseExpr*(s: string): NimNode {.noSideEffect, compileTime.} =
...
+macro strJoin*(args: varargs[typed]): untyped =
+  ## same as `$args[0] & ... & $args[^1]` but in lots of cases more
+  ## convenient, to avoid visual noise from lots of operators & and $.
+  ## Also, makes it easier to switch bewtween `echo` and `strJoin`.
+  runnableExamples:
+    doAssert strJoin() == ""
+    ## easier to read than: `$(1+2) & " " & $(3-4)`
+    doAssert strJoin(1+2, " ", 4-3) == "3 1"
+  if args.len > 0:
+    var cmd = ""
```
Comment (Araq) [inline lib/core/macros.nim]: Terrible way to construct an AST.

**Comment by Araq** [comment]
-1 from me. The stdlib is not a dumping ground for every helper that's useful to somebody. That's how we got into the current mess...

File: lib/core/macros.nim (hunk 1 above)
Comment (timotheecour) [reply lib/core/macros.nim]: fixed with latest commit.

**Comment by timotheecour** [comment]
> And move it to strformat?

it would indeed be more logical there but I've seen before `strformat` being avoided because it'd add too much stuff, and I wanted this to be importable with minimum import dependencies (see my comment above)

> -1 from me. The stdlib is not a dumping ground for every helper that's useful to somebody. That's how we got into the current mess...

I agree with this sentiment in general but stringified concat is used very frequently and strJoin makes the code easier to read/write/convert from echo.

I found a way to avoid `parseExpr` while remaining as efficient as `expr & expr `.
The trick was to construct an Ast comprised of a single root expression instead of what I had a few commits ago which was based on multiple `add` statements, which prevented the efficient term rewriting.

Actually @Araq I couldnt' find where this term rewriting happens in the code (ie where is the term rewriting rule for `a & b & c...`) ?

**Comment by tim-st** [comment]
I'd like to have this working in `strutils` but with an optional seperator `sep = ""`. But when there is a seperator the syntax isn't clear anymore unless `sep = "..."` is always named explicitly...
If the seperator should always be `""` like in the example, the following would work:

```nim
import strutils

proc join(a: varargs[string, `$`]): string =
  result = join(a, "")

echo join(1, 1 == 2, "ab", 'c', @[1, 2, 3], 4-3)
```
